File: 2022/deno/09.ts
const input = await Deno.readTextFile("./input/09.in");

const data = input.split("\n");

/**
 * Part One
 */

const moves = data
  .map((row) => row.split(" "))
  .map(([key, value]) => [key, Number(value)] as const);

const asKey = (coord: number[]) => `${coord[0]}.${coord[1]}`;

const distance = (head: number[], tail: number[]) => {
  const delta = [head[0] - tail[0], head[1] - tail[1]];

  return Math.abs(delta[0]) + Math.abs(delta[1]);
};

const calcMove = (head: number[], tail: number[]) => {
  const delta = [head[0] - tail[0], head[1] - tail[1]];
  const dx = delta[0];
  const dy = delta[1];

  return [dx === 0 ? dx : Math.sign(dx), dy === 0 ? dy : Math.sign(dy)];
};

const isTouching = (head: number[], tail: number[]) => {
  const dist = distance(head, tail);

  if (dist < 0) throw new Error("Negative distance");

  if (dist > 2) return false;

  if (dist === 2) {
    const delta = [head[0] - tail[0], head[1] - tail[1]];

    // same x
    if (delta[0] === 0) return false;
    // same y
    if (delta[1] === 0) return false;
  }

  return true;
};

const countUniqueTailPositions = (rope: number[][]) => {
  // x, y
  const tail = rope[rope.length - 1];
  const [head, ...rest] = rope;

  const tailSteps = new Set<string>();
  tailSteps.add(asKey(tail));

  moves.forEach((move) => {
    const [dir, steps] = move;

    for (const _ of Array.from({ length: steps })) {
      switch (dir) {
        case "U":
          head[1] = head[1] + 1;
          break;
        case "D":
          head[1] = head[1] - 1;
          break;
        case "L":
          head[0] = head[0] - 1;
          break;
        case "R":
          head[0] = head[0] + 1;
          break;
      }

      while (
        rest.some((curr, index) => {
          return !isTouching(rope[index], curr);
        })
      ) {
        rest.forEach((curr, index) => {
          const lead = rope[index];

          if (!isTouching(lead, curr)) {
            const [dx, dy] = calcMove(lead, curr);
            curr[0] = curr[0] + dx;
            curr[1] = curr[1] + dy;

            if (curr === tail) {
              tailSteps.add(asKey(curr));
            }
          }
        });
      }
    }
  });

  return tailSteps.size;
};

console.log(
  "Part one:",
  countUniqueTailPositions([
    [0, 0],
    [0, 0],
  ])
);

/**
 * Part Two
 */

const longRope = Array.from({ length: 10 }, () => [0, 0]);

console.log("Part two:", countUniqueTailPositions(longRope));

File: 2022/swift/09.swift
import Foundation

let filename = "./input/09.in"

typealias Point = (x: Int, y: Int)

func sign(_ num: Int) -> Int {
    num > 0 ? 1 : -1
}

func dist(_ lhs: Point, _ rhs: Point) -> Int {
    abs(rhs.x - lhs.x) + abs(rhs.y - lhs.y)
}

func touching(_ lhs: Point, _ rhs: Point) -> Bool {
    let vec = [0, 1, -1]

    return vec.first {
        dx in
        vec.first { dy in
            (lhs.x + dx) == rhs.x && (lhs.y + dy) == rhs.y
        } != nil
    } != nil
}

func follow(_ current: Point, _ lead: Point) -> Point {
    // same col
    if lead.x == current.x {
        if dist(lead, current) >= 2 {
            if lead.y < current.y {
                // lead is above
                return (x: 0, y: -1)
            }
            return (x: 0, y: 1)
        }
    }

    // same row
    if lead.y == current.y {
        if dist(lead, current) >= 2 {
            if lead.x < current.x {
                return (x: -1, y: 0)
            }
            return (x: 1, y: 0)
        }
    }

    if !touching(lead, current) {
        let dx = sign(lead.x - current.x)
        let dy = sign(lead.y - current.y)

        return (x: dx, y: dy)
    }

    return (x: 0, y: 0)
}

class Knot {
    var point: Point {
        didSet {
            coverage.insert("\(point.x).\(point.y)")
        }
    }

    var coverage: Set<String>

    init(_ point: Point) {
        self.point = point
        coverage = Set<String>()
        coverage.insert("\(point.x).\(point.y)")
    }

    func move(_ by: Point) {
        let (dx, dy) = by
        point = (point.x + dx, point.y + dy)
    }
}

typealias Simulate = ([Knot]) -> Void

func main() {
    do {
        let contents = try String(contentsOfFile: filename)

        let instructions = contents.split(separator: "\n").map {
            row in
            let entry = row.split(separator: " ")
            let delta = Int(entry[1])!

            switch entry[0] {
            case "L":
                return (x: -delta, y: 0)
            case "R":
                return (x: delta, y: 0)
            case "U":
                return (x: 0, y: -delta)
            case "D":
                return (x: 0, y: delta)
            default:
                return (0, 0)
            }
        }

        let simulate: Simulate = {
            rope in

            for inst in instructions {
                rope[0].move(inst)

                let body = rope.suffix(rope.count - 1).enumerated()

                while !body.allSatisfy({ touching($0.1.point, rope[$0.0].point) }) {
                    for (index, knot) in body {
                        knot.move(follow(knot.point, rope[index].point))
                    }
                }
            }
        }

        let head = Knot((x: 0, y: 0))
        let tail = Knot((x: 0, y: 0))

        let rope = [head, tail]

        simulate(rope)

        print("Part one:", tail.coverage.count)

        let longRope = Array(0 ... 9).map { _ in Knot((x: 0, y: 0)) }

        simulate(longRope)

        print("Part two:", longRope[longRope.count - 1].coverage.count)

    } catch {
        print(error)
    }
}

main()
